pkg/apis/navigator/validation/elasticsearch.go

```diff
@@ -0,0 +1,107 @@
+package validation
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	apimachineryvalidation "k8s.io/apimachinery/pkg/api/validation"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"github.com/jetstack/navigator/pkg/apis/navigator"
+)
+
+var supportedPullPolicies = []string{
+	string(corev1.PullNever),
+	string(corev1.PullIfNotPresent),
+	string(corev1.PullAlways),
+	"",
+}
+
+func ValidateImageSpec(img *navigator.ImageSpec, fldPath *field.Path) field.ErrorList {
+	el := field.ErrorList{}
+	if img.Tag == "" {
+		el = append(el, field.Required(fldPath.Child("tag"), ""))
+	}
+	if img.Repository == "" {
+		el = append(el, field.Required(fldPath.Child("repository"), ""))
+	}
+	if img.PullPolicy != corev1.PullNever &&
+		img.PullPolicy != corev1.PullIfNotPresent &&
+		img.PullPolicy != corev1.PullAlways &&
+		img.PullPolicy != "" {
+		el = append(el, field.NotSupported(fldPath.Child("pullPolicy"), img.PullPolicy, supportedPullPolicies))
+	}
+	return el
+}
+
+var supportedESClusterRoles = []string{
+	string(navigator.ElasticsearchRoleData),
+	string(navigator.ElasticsearchRoleIngest),
+	string(navigator.ElasticsearchRoleMaster),
+}
+
+func ValidateElasticsearchClusterRole(r navigator.ElasticsearchClusterRole, fldPath *field.Path) field.ErrorList {
+	el := field.ErrorList{}
+	switch r {
+	case navigator.ElasticsearchRoleData:
+	case navigator.ElasticsearchRoleIngest:
+	case navigator.ElasticsearchRoleMaster:
+	default:
+		el = append(el, field.NotSupported(fldPath, r, supportedESClusterRoles))
+	}
+	return el
+}
+
+func ValidateElasticsearchClusterNodePool(np *navigator.ElasticsearchClusterNodePool, fldPath *field.Path) field.ErrorList {
+	el := ValidateDNS1123Subdomain(np.Name, fldPath.Child("name"))
+	el = append(el, ValidateElasticsearchPersistence(&np.Persistence, fldPath.Child("persistence"))...)
+	rolesPath := fldPath.Child("roles")
+	if len(np.Roles) == 0 {
+		el = append(el, field.Required(rolesPath, "at least one role must be specified"))
+	}
+	for i, r := range np.Roles {
+		idxPath := rolesPath.Index(i)
+		el = append(el, ValidateElasticsearchClusterRole(r, idxPath)...)
+	}
+	if np.Replicas < 0 {
+		el = append(el, field.Invalid(fldPath.Child("replicas"), np.Replicas, "must be greater than zero"))
+	}
+	// TODO: call k8s.io/kubernetes/pkg/apis/core/validation.ValidateResourceRequirements on np.Resources
+	// this will require vendoring kubernetes/kubernetes and switching to use the corev1 ResourceRequirements
+	// struct
+	return el
+}
+
+func ValidateElasticsearchPersistence(cfg *navigator.ElasticsearchClusterPersistenceConfig, fldPath *field.Path) field.ErrorList {
+	el := field.ErrorList{}
+	if cfg.Enabled && cfg.Size.IsZero() {
+		el = append(el, field.Required(fldPath.Child("size"), ""))
+	}
+	if cfg.Size.Sign() == -1 {
+		el = append(el, field.Invalid(fldPath.Child("size"), cfg.Size, "must be greater than zero"))
+	}
```

Comment:
Perhaps instead check that the value is >0 and this could all be replaced with, `if cgf.Size.Sign() != 1`

The docs say:

```
// Sign returns 0 if the quantity is zero, -1 if the quantity is less than zero, or 1 if the
// quantity is greater than zero.

```

Reply:
As in my other comment, there are a few cases to cover:

1) Persistence enabled, size is >0 (VALID)
3) Persistence enabled, size = 0 (INVALID)
2) Persistence disabled, size = 0 (VALID)
4) Size<0 - ALWAYS invalid

I've expanded our test suite to explicitly cover all the cases above, and they are still passing 😄

Reply:
I still don't see the case for `3 Persistence disabled, size = 0 (VALID)`
Unless that's how we determine that persistence is not required.
In which case, would that be better expressed as `Persistence  *ElasticsearchClusterPersistenceConfig` and `Persistence = nil` ?

Reply:
Ah, I see you just explained below.

Reply:
So if I set:

```persistence:
  enabled: false```

I have implicitly also said the size=0. It would we weird/awkward to require users to write:

```persistence:
  enabled: false
  # size is ignored
  size: 1Gi
```